aether-ui/aether/ui/api/models.py

```diff
@@ -23,19 +23,40 @@
 from django_prometheus.models import ExportModelOperationsMixin
 from model_utils.models import TimeStampedModel
 
-from .utils import validate_pipeline
+from .utils import validate_contract
 
 
 class Pipeline(ExportModelOperationsMixin('ui_pipeline'), TimeStampedModel):
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
     name = models.CharField(max_length=100, null=False, blank=False, unique=True)
-
     # this is the avro schema
     schema = JSONField(blank=True, null=True, default={})
 
     # this is an example of the data using the avro schema
     input = JSONField(blank=True, null=True, default={})
 
+    # this is a reference to the linked kernel mappingset
+    mappingset = models.UUIDField(null=True, blank=True)
+
+    def __str__(self):
+        return self.name
+
+    def save(self, *args, **kwargs):
+        read_only_contracts = self.contracts.filter(is_read_only=True)
+        if read_only_contracts:
+            raise Exception('Input is readonly for this pipeline')
```

Comment:
Why are you raising this error? Shouldn't you check first that you are changing the input? What happens if you are just changing the `mappingset` value?

Reply:
I assume a readonly mapping sets means exactly that. All its properties included. In the case of the ui, same is true for the pipeline if any of its nested contracts are marked as readonly, and changes should only be made on the contract level. So as not to cause any form of regression to any apps dependent of the mapping set.